### ISSUE TITLE
WARNING in alloc_netdev_mqs

### ISSUE THREAD

**Issue by daxtens** [linuxppc/issues#284]
Per https://syzkaller-ppc64.appspot.com/bug?id=cfdf75cd985012d0124cd41e6fa095d33e7d0f6b

I can reproduce it with a kernel compiled with the syzkaller config from @ajdlinux, I just have to make sure that `panic_on_warn` is set.

```
[   75.520736] Kernel panic - not syncing: panic_on_warn set ...
[   75.520859] CPU: 0 PID: 5698 Comm: syz-executor Not tainted 5.4.0-rc7-next-20191118 #187
[   75.520929] Call Trace:
[   75.520977] [c00000000840b210] [c000000001865ad4] dump_stack+0x190/0x25c (unreliable)
[   75.521054] [c00000000840b280] [c0000000001cb134] panic+0x268/0x734
[   75.521117] [c00000000840b320] [c0000000001cadc0] __warn+0x160/0x1f0
[   75.521180] [c00000000840b3b0] [c0000000018638bc] report_bug+0x1cc/0x360
[   75.521245] [c00000000840b450] [c000000000043510] program_check_exception+0x4b0/0x6f0
[   75.521320] [c00000000840b4e0] [c0000000000091dc] program_check_common+0x26c/0x270
[   75.521397] --- interrupt: 700 at lockdep_register_key+0x1b0/0x1d0
[   75.521397]     LR = lockdep_register_key+0x20/0x1d0
[   75.521492] [c00000000840b7e0] [c00000000840b850] 0xc00000000840b850 (unreliable)
[   75.521577] [c00000000840b850] [c00000000134cbcc] alloc_netdev_mqs+0x19c/0x700
[   75.521659] [c00000000840b900] [c000000000fe8268] loopback_net_init+0x58/0x160
[   75.521732] [c00000000840b980] [c000000001336f84] ops_init+0xa4/0x350
[   75.521818] [c00000000840b9d0] [c000000001337368] setup_net+0x138/0x380
[   75.521882] [c00000000840ba50] [c000000001338738] copy_net_ns+0x198/0x430
[   75.521946] [c00000000840baa0] [c00000000022ee3c] create_new_namespaces+0x3ac/0x440
[   75.522019] [c00000000840baf0] [c00000000022f2b8] copy_namespaces+0x3e8/0x440
[   75.522092] [c00000000840bb50] [c0000000001c6340] copy_process+0x1250/0x28e0
[   75.522166] [c00000000840bc90] [c0000000001c7d9c] _do_fork+0xec/0xf50
[   75.522229] [c00000000840bd60] [c0000000001c90cc] sys_clone+0xdc/0x140
[   75.522292] [c00000000840be20] [c00000000000bed8] ppc_clone+0x8/0xc
```

I'm narrowing down the syzkaller reproducer.

**Comment by daxtens** [comment]
This is a shrunk reproducer, I'm still minimising it. It needs to be run repeatedly (`-repeat 10000`)  to leak all the memory, the warn happens (AFAICT) after the first oomkill so it's probably an error path not being handled correctly.

```
# https://syzkaller-ppc64.appspot.com/bug?id=cfdf75cd985012d0124cd41e6fa095d33e7d0f6b
# See https://goo.gl/kgGztJ for information about syzkaller reproducers.
#{"threaded":true,"repeat":true,"procs":1,"sandbox":"","fault_call":-1,"close_fds":false}
r4 = socket$inet_udp(0x2, 0x2, 0x0)
connect$inet(r4, &(0x7f0000000340)={0x2, 0x0, @remote}, 0x10)
fcntl$dupfd(r4, 0x0, r4)
setsockopt$IPT_SO_SET_ADD_COUNTERS(0xffffffffffffffff, 0x0, 0x41, 0x0, 0x0)
connect$inet(0xffffffffffffffff, 0x0, 0x0)
fcntl$dupfd(0xffffffffffffffff, 0x0, 0xffffffffffffffff)
clone(0x60200000, 0x0, 0x0, 0x0, 0x0)
```

**Comment by daxtens** [comment]
Here's an edited down C reproducer, turns out all you need do is call `clone()` a lot, which is distressing:

```
// autogenerated by syzkaller (https://github.com/google/syzkaller) and trimmed

#define __unix__ 1
#define __gnu_linux__ 1
#define __linux__ 1

#define _GNU_SOURCE

#include <dirent.h>
#include <endian.h>
#include <errno.h>
#include <fcntl.h>
#include <signal.h>
#include <stdarg.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/prctl.h>
#include <sys/stat.h>
#include <sys/syscall.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <time.h>
#include <unistd.h>

static uint64_t current_time_ms(void)
{
	struct timespec ts;
	if (clock_gettime(CLOCK_MONOTONIC, &ts))
		exit(1);
	return (uint64_t)ts.tv_sec * 1000 + (uint64_t)ts.tv_nsec / 1000000;
}

static void kill_and_wait(int pid, int* status)
{
	kill(-pid, SIGKILL);
	kill(pid, SIGKILL);
	int i;
	for (i = 0; i < 100; i++) {
		if (waitpid(-1, status, WNOHANG | __WALL) == pid)
			return;
		usleep(1000);
	}
	while (waitpid(-1, status, __WALL) != pid) {
	}
}

#ifndef __NR_clone
#define __NR_clone 120
#endif


int main(void)
{
	while (true) {
		int pid = fork();
		if (pid < 0)
			exit(1);
		if (pid == 0) {
			syscall(__NR_clone, 0x60200000ul, 0ul, 0ul, 0ul, 0ul);
			exit(0);
		}
		int status = 0;
		uint64_t start = current_time_ms();
		for (;;) {
			if (waitpid(-1, &status, WNOHANG | __WALL) == pid)
				break;
			usleep(1000);
			if (current_time_ms() - start < 5 * 1000)
				continue;
			kill_and_wait(pid, &status);
			break;
		}
	}
	return 0;
}
```

It will panic after the first OOMkill, AFAICT.

**Comment by daxtens** [comment]
The shrunk syz reproducer is:
```
# https://syzkaller-ppc64.appspot.com/bug?id=cfdf75cd985012d0124cd41e6fa095d33e7d0f6b
# See https://goo.gl/kgGztJ for information about syzkaller reproducers.
#{"threaded":true,"repeat":true,"procs":1,"sandbox":"","fault_call":-1,"close_fds":false}
clone(0x60200000, 0x0, 0x0, 0x0, 0x0)
```

**Comment by mpe** [comment]
There's two `WARN_ON()`'s in `lockdep_register_key()`, which one are we hitting?

**Comment by daxtens** [comment]
Syzkaller reports `WARNING: CPU: 0 PID: 6403 at kernel/locking/lockdep.c:1120 ...`:

```
	if (WARN_ON_ONCE(static_obj(key)))
		return;
```

I'm bisecting it atm, it looks like it's limited to -next.

**Comment by daxtens** [comment]
Bisecting, this is `ab92d68fc22f ("net: core: add generic lockdep keys")` which already has a slew of syzkaller reports/bisections.

**Comment by mpe** [comment]
Could be another case of:

https://lore.kernel.org/linuxppc-dev/87ef0vpfbc.fsf@mpe.ellerman.id.au/

**Comment by daxtens** [comment]
Yes, it is. Good spot, thanks!

**Comment by mpe** [comment]
Should be fixed by torvalds/linux@6f07048c00fd100ed8cab66c225c157e0b6c0a50.
